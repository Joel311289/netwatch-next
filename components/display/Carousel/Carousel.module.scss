/** Carousel (main component) **/

.wrapper {
  --swiper-pagination-color: var(--color-primary);

  box-sizing: border-box;
  position: relative;
}

.buttonNavigation {
  color: var(--color-button-text);
  background-color: var(--color-dark);
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  border-radius: 5px;
  z-index: 9;
  height: 50px !important;
  padding: 5px 0 !important;
  font-size: 2rem !important;
  min-width: 25px;
  font-weight: 700;
  opacity: 0.35;
  display: flex;
  justify-content: center;

  &:hover {
    opacity: 0.7 !important;
  }

  &.prev {
    left: 5px;
  }

  &.next {
    right: 5px;
  }

  .xs & {
    display: none;
  }
}

.buttonNavigationDisabled {
  display: none !important;
}

.paginationBulletsClass {
  width: 100% !important;
  display: flex;
  justify-content: center;
  margin-bottom: 0.5em;

  .xs &,
  .sm & {
    bottom: 0;
    margin: 1px;
  }
}

.paginationBulletClass {
  background-color: var(--color-disabled);
  opacity: 0.5;
  border-radius: 0;
  width: 20px;
  height: 4px;
  margin: 0 2px;
  cursor: pointer;

  .xs &,
  .sm & {
    flex-grow: 1;
    margin: 0;
    background-color: var(--color-primary);
    opacity: 1;
    transition: all 0.3s ease-in;
  }

  .xs .paginationBulletActiveClass ~ &,
  .sm .paginationBulletActiveClass ~ & {
    background-color: var(--color-disabled) !important;
    opacity: 0.5 !important;
  }
}

.paginationBulletActiveClass {
  opacity: 1;
  background-color: var(--color-primary);
}

/** CarouselItem (child component) **/

.item {
  width: 300px;
}
