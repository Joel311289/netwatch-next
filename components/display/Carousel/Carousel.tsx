import { createContext, useContext, cloneElement } from 'react';
import { Swiper, SwiperSlide } from 'swiper/react';
import SwiperCore, {
  Autoplay,
  Mousewheel,
  Navigation,
  Pagination,
  EffectFade,
  SwiperOptions
} from 'swiper';
import { IconType } from 'react-icons';
import { FiChevronLeft, FiChevronRight } from 'react-icons/fi';
import { ElementChildren, ElementHTML } from '@/types';
import { withChildrenFiltered } from '@/hoc/withChildrenFiltered';
import { Button } from '@/components/forms';
import { classes } from '@/utils/helpers';
import styles from '@/components/display/Carousel/Carousel.module.scss';

SwiperCore.use([EffectFade, Autoplay, Mousewheel, Navigation, Pagination]);

/* -------------------------------------------------------------------------- */
/** Context **/
/* -------------------------------------------------------------------------- */

type CarouselContextProps = typeof defaultValue;

const defaultValue = {};

const CarouselContext = createContext<CarouselContextProps>(defaultValue);

const useCarouselContext = () => useContext(CarouselContext);

/* -------------------------------------------------------------------------- */
/** CarouselItem (child component) **/
/* -------------------------------------------------------------------------- */

type CarouselItemProps = ElementHTML & ElementChildren<JSX.Element>;

const CarouselItem = ({ children, className }: CarouselItemProps) => {
  useCarouselContext();

  return <div className={classes(styles.item, className)}>{children}</div>;
};

/* -------------------------------------------------------------------------- */
/** Carousel (main component) **/
/* -------------------------------------------------------------------------- */

export type CarouselProps = typeof defaultProps &
  ElementHTML &
  ElementChildren & {
    navigation?: boolean;
    pagination?: boolean;
    autoplay?: boolean;
    slidesPerView?: number | 'auto';
    spacing?: number;
    effectFade?: boolean;
  };

const defaultProps = {
  children: [],
  spacing: 2
};

const renderNavigationButton = (state: 'prev' | 'next', Icon: IconType) => {
  return (
    <Button className={classes(styles.buttonNavigation, styles[state])}>
      <Icon />
    </Button>
  );
};

const Carousel = ({
  className,
  children,
  navigation,
  slidesPerView,
  spacing,
  autoplay,
  effectFade
}: CarouselProps) => {
  const settings: SwiperOptions = {
    speed: 500,
    allowTouchMove: false,
    slidesPerView,
    slidesPerGroup: slidesPerView === 'auto' ? 1 : slidesPerView,
    spaceBetween: spacing,
    navigation: navigation
      ? {
          prevEl: `.${styles.prev}`,
          nextEl: `.${styles.next}`,
          disabledClass: styles.buttonNavigationDisabled
        }
      : {},
    autoplay: Boolean(autoplay) && {
      delay: 5000,
      pauseOnMouseEnter: true
    },
    ...(effectFade ? { effect: 'fade' } : {}),
    watchSlidesProgress: true,
    mousewheel: {
      forceToAxis: true
    }
  };

  return (
    <CarouselContext.Provider value={{}}>
      <Swiper role="list" {...settings} className={classes(styles.wrapper, className)}>
        {navigation && renderNavigationButton('prev', FiChevronLeft)}
        {navigation && renderNavigationButton('next', FiChevronRight)}
        {children.map((element: JSX.Element, index) => (
          <SwiperSlide key={index}>{cloneElement(element, { ...element.props })}</SwiperSlide>
        ))}
      </Swiper>
    </CarouselContext.Provider>
  );
};

Carousel.defaultProps = defaultProps;

const CarouselWithChildrenFiltered = withChildrenFiltered<CarouselProps>(Carousel, {
  Item: CarouselItem
});

export default CarouselWithChildrenFiltered;
