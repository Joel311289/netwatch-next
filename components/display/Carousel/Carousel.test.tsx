import { render, screen } from '@testing-library/react';
import Carousel from '@/components/display/Carousel/Carousel';

const items = [...Array(3)].map((_, index) => (
  <Carousel.Item key={index}>item ${index + 1}</Carousel.Item>
));

describe('Tests Carousel component', () => {
  test('renders component correctly', () => {
    const { container } = render(<Carousel>{items}</Carousel>);

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    const spacing = 5;
    render(
      <Carousel spacing={spacing} navigation>
        {items}
      </Carousel>
    );

    const slides = screen.getByRole('list').childNodes[0];

    expect(slides.childNodes.length).toBe(items.length);
  });

  test('renders no children correctly', () => {
    const spacing = [5, 10];
    render(<Carousel spacing={spacing} navigation />);

    const slides = screen.getByRole('list').childNodes[0];

    expect(slides.childNodes.length).toBe(0);
  });
});
