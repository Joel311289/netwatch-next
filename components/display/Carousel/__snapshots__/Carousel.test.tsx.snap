// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Carousel component renders component correctly 1`] = `
<div>
  <div
    class="swiper-container swiper-container-initialized swiper-container-horizontal wrapper md"
    role="list"
    style="width: calc(100% + 2 * 0px); margin-left: 0px;"
  >
    <div
      class="swiper-wrapper"
      style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
    >
      <div
        class="swiper-slide item"
      >
        <div
          class=""
        >
          item $
          1
        </div>
      </div>
      <div
        class="swiper-slide item"
      >
        <div
          class=""
        >
          item $
          2
        </div>
      </div>
      <div
        class="swiper-slide item"
      >
        <div
          class=""
        >
          item $
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;
